matrix_sdk_appservice/src/lib.rs

```diff
@@ -217,26 +203,21 @@ impl Appservice {
     ) -> Result<Self> {
         let homeserver_url = homeserver_url.try_into()?;
         let server_name = server_name.try_into()?;
+        let registration = Arc::new(registration);
+        let clients = HashMap::new();
 
-        let client_sender_localpart = Client::new(homeserver_url.clone())?;
-
-        client_session_with_login_restore(
-            &client_sender_localpart,
-            &registration,
-            registration.sender_localpart.as_ref(),
-            &server_name,
-        )
-        .await?;
-
-        Ok(Appservice { homeserver_url, server_name, registration, client_sender_localpart })
+        Ok(Appservice { homeserver_url, server_name, registration, clients })
     }
 
     /// Get a [`Client`]
     ///
     /// Will return a `Client` that's configured to [assert the identity] on all
     /// outgoing homeserver requests if `localpart` is given. If not given
     /// the `Client` will use the main user associated with this appservice,
-    /// that is the `sender_localpart` in the [`AppserviceRegistration`]
+    /// that is the `sender_localpart` in the [`AppserviceRegistration`].
+    ///
+    /// This method is a singleton that safes the client internally for re-use
```

Comment:
Does it save the client internally? It's probably supposed to store it in the `self.clients` `HashMap` but that would require a mutable borrow, so it can't be doing it?

Reply:
Thanks for the catch, I thought that I was doing it already, but haven't actually finished it last week :facepalm: Should be fine now